mlir/lib/Conversion/MQTRefToQIR/MQTRefToQIR.cpp

```diff
@@ -0,0 +1,881 @@
+        op.getLoc(), fnDecl,
+        ValueRange{resultValue, addressOfOp->getResult(0)});
+
+    // erase the old operation
+    rewriter.eraseOp(op);
+
+    // increase the number of required results
+    getState().numResults++;
+
+    return success();
+  }
+};
+
+struct MQTRefToQIR final : impl::MQTRefToQIRBase<MQTRefToQIR> {
+  using MQTRefToQIRBase::MQTRefToQIRBase;
+
+  /**
+   * @brief Finds the main function in the module
+   *
+   * @param op The module operation that holds all operations.
+   * @return The main function.
+   */
+  static LLVM::LLVMFuncOp getMainFunction(Operation* op) {
+    auto module = dyn_cast<ModuleOp>(op);
+    // find the main function
+    for (auto funcOp : module.getOps<LLVM::LLVMFuncOp>()) {
+      auto passthrough = funcOp->getAttrOfType<ArrayAttr>("passthrough");
+      if (!passthrough) {
+        continue;
+      }
+      if (llvm::any_of(passthrough, [](Attribute attr) {
+            auto strAttr = dyn_cast<StringAttr>(attr);
+            return strAttr && strAttr.getValue() == "entry_point";
+          })) {
+        return funcOp;
+      }
+    }
+    return nullptr;
+  }
+
+  /**
+   * @brief Makes sure that the different blocks for the base profile of QIR
+   * exist
+   *
+   * The first block should only contain constant operations for the initialize
+   * operation. The second block contains the reversible quantum operations. The
+   * third block contains the irreversible quantum operations like the measure
+   * operation or the dealloc operation. The final block contains the
+   * return operation and the record output operation. The blocks are connected
+   * with an unconditional jump operation to the next block.
+   *
+   * @param main The main function of the module.
+   */
+  static void ensureBlocks(LLVM::LLVMFuncOp& main) {
+    // return if there are more blocks already
+    if (main.getBlocks().size() > 1) {
+      return;
+    }
+    // get the existing block
+    auto* entryBlock = &main.front();
+    OpBuilder builder(main.getBody());
+
+    // create the main blocks and the endblock
+    Block* mainBlock = builder.createBlock(&main.getBody());
+    Block* irreversibleMainBlock = builder.createBlock(&main.getBody());
+    Block* endBlock = builder.createBlock(&main.getBody());
+
+    // move the returnOp from the entryBlock to the endBlock
+    auto& entryOperations = entryBlock->getOperations();
+    auto& endOperations = endBlock->getOperations();
+    endOperations.splice(endOperations.end(), entryOperations,
+                         std::prev(entryOperations.end()));
+
+    // get all quantum operations until the measure operation
+    SmallVector<Operation*> opsToMove;
+    for (auto& op : entryOperations) {
+      if (dyn_cast<ref::MeasureOp>(op)) {
+        break;
+      }
+      opsToMove.emplace_back(&op);
+    }
+    // move them to the 2nd block
+    for (Operation* op : opsToMove) {
+      op->moveBefore(mainBlock, mainBlock->end());
+    }
```

Comment:
Why construct this `opsToMove` vector at all? Can't you do this in-place somehow during the iteration? Seems kind of wasteful to me.

Another thought: given how, especially in base profile programs, the majority of operations will be quantum operations, why not simply keep the quantum operations in the main block and inserting the other blocks around that?